src/storage/v2/inmemory/edge_type_property_index.cpp

```diff
@@ -0,0 +1,332 @@
+// License, and you may not use this file except in compliance with the Business Source License.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+#include "storage/v2/inmemory/edge_type_property_index.hpp"
+
+#include "storage/v2/constraints/constraints.hpp"
+#include "storage/v2/edge_info_helpers.hpp"
+#include "storage/v2/id_types.hpp"
+#include "storage/v2/indices/indices_utils.hpp"
+#include "utils/counter.hpp"
+
+namespace {
+
+using Delta = memgraph::storage::Delta;
+using Vertex = memgraph::storage::Vertex;
+using Edge = memgraph::storage::Edge;
+using EdgeRef = memgraph::storage::EdgeRef;
+using EdgeTypeId = memgraph::storage::EdgeTypeId;
+using PropertyId = memgraph::storage::PropertyId;
+using PropertyValue = memgraph::storage::PropertyValue;
+using Transaction = memgraph::storage::Transaction;
+using View = memgraph::storage::View;
+
+using ReturnType = std::optional<std::tuple<EdgeTypeId, Vertex *, EdgeRef>>;
+ReturnType VertexDeletedConnectedEdges(Vertex *vertex, Edge *edge, const Transaction *transaction, View view) {
+  ReturnType link;
+  Delta *delta = nullptr;
+  {
+    auto guard = std::shared_lock{vertex->lock};
+    delta = vertex->delta;
+  }
+  ApplyDeltasForRead(transaction, delta, view, [&edge, &vertex, &link](const Delta &delta) {
+    switch (delta.action) {
+      case Delta::Action::ADD_LABEL:
+      case Delta::Action::REMOVE_LABEL:
+      case Delta::Action::SET_PROPERTY:
+        break;
+      case Delta::Action::ADD_IN_EDGE: {
+        if (edge == delta.vertex_edge.edge.ptr) {
+          link = {delta.vertex_edge.edge_type, delta.vertex_edge.vertex, delta.vertex_edge.edge};
+          auto it = std::find(vertex->in_edges.begin(), vertex->in_edges.end(), link);
+          MG_ASSERT(it == vertex->in_edges.end(), "Invalid database state!");
+          break;
+        }
+      }
+      case Delta::Action::ADD_OUT_EDGE: {
+        if (edge == delta.vertex_edge.edge.ptr) {
+          link = {delta.vertex_edge.edge_type, delta.vertex_edge.vertex, delta.vertex_edge.edge};
+          auto it = std::find(vertex->out_edges.begin(), vertex->out_edges.end(), link);
+          MG_ASSERT(it == vertex->out_edges.end(), "Invalid database state!");
+          break;
+        }
+      }
+      case Delta::Action::REMOVE_IN_EDGE:
+      case Delta::Action::REMOVE_OUT_EDGE:
+      case Delta::Action::RECREATE_OBJECT:
+      case Delta::Action::DELETE_DESERIALIZED_OBJECT:
+      case Delta::Action::DELETE_OBJECT:
+        break;
+    }
+  });
+  return link;
+}
+}  // namespace
+
+namespace memgraph::storage {
+
+bool InMemoryEdgeTypePropertyIndex::Entry::operator<(const Entry &rhs) const {
+  if (edge->gid < rhs.edge->gid) {
+    return true;
+  }
+  if (rhs.edge->gid < edge->gid) {
+    return false;
+  }
```

Comment:
We already checked `edge->gid < rhs.edge->gid` in previous comparison. But I think this can be rewritten as `edge->gid > rhs.edge->gid` to be more clear.
In the case of edge gids equality we will compare tuples.